src/audio/pcm_converter/pcm_converter_hifi3.c

```diff
@@ -50,79 +40,50 @@ static int pcm_convert_s16_to_s24(const struct audio_stream __sparse_cache *sour
 				  uint32_t ioffset, struct audio_stream __sparse_cache *sink,
 				  uint32_t ooffset, uint32_t samples)
 {
-	ae_int16 *in = audio_stream_read_frag(source, ioffset, sizeof(int16_t));
-	ae_int32 *out = audio_stream_write_frag(sink, ooffset, sizeof(int32_t));
 	ae_int16x4 sample = AE_ZERO16();
-	ae_valign align_out = AE_ZALIGN64();
-	ae_int16x4 *in16x4;
-	ae_int32x2 *out32x2;
-	int i = 0;
-
-	/* nothing to do */
-	if (!samples)
-		return samples;
-
-	/* required alignment for AE_L16X4_XC */
-	while (!IS_ALIGNED((uintptr_t)in, 8)) {
-		/* set source as circular buffer */
-		pcm_converter_setup_circular(source);
-
-		/* load one 16 bit sample */
-		AE_L16_XC(sample, in, sizeof(ae_int16));
-
-		/* set sink as circular buffer */
-		pcm_converter_setup_circular(sink);
-
-		/* shift right and store one 32 bit sample */
-		AE_S32_L_XC(AE_SRAI32(AE_CVT32X2F16_32(sample), 8), out,
-			    sizeof(ae_int32));
-
-		if (++i == samples)
-			return samples;
-	}
-
-	in16x4 = (ae_int16x4 *)in;
-	out32x2 = (ae_int32x2 *)out;
-
-	/* main loop processes 4 samples at a time */
-	while (samples >= 3 && i < samples - 3) {
-		/* set source as circular buffer */
-		pcm_converter_setup_circular(source);
-
-		/* load four 16 bit samples */
-		AE_L16X4_XC(sample, in16x4, sizeof(ae_int16x4));
-
-		/* set sink as circular buffer */
-		pcm_converter_setup_circular(sink);
-
-		/* shift right and store four 32 bit samples */
-		AE_SA32X2_IC(AE_SRAI32(AE_CVT32X2F16_32(sample), 8), align_out,
-			     out32x2);
-		AE_SA32X2_IC(AE_SRAI32(AE_CVT32X2F16_10(sample), 8), align_out,
-			     out32x2);
-
-		i += 4;
-	}
-
-	/* flush align_out register to memory */
-	AE_SA64POS_FC(align_out, out32x2);
-
-	in = (ae_int16 *)in16x4;
-	out = (ae_int32 *)out32x2;
-
-	while (i++ != samples) {
-		/* set source as circular buffer */
-		pcm_converter_setup_circular(source);
-
-		/* load one 16 bit sample */
-		AE_L16_XC(sample, in, sizeof(ae_int16));
-
-		/* set sink as circular buffer */
-		pcm_converter_setup_circular(sink);
-
-		/* shift right and store one 32 bit sample */
-		AE_S32_L_XC(AE_SRAI32(AE_CVT32X2F16_32(sample), 8), out,
-			    sizeof(ae_int32));
+	int nmax, i, n, m, left;
+	ae_valign inu = AE_ZALIGN64();
+	ae_valign outu = AE_ZALIGN64();
+	int32_t left_samples = samples;
+	int16_t *src = source->r_ptr;
+	int32_t *dst = sink->w_ptr;
+
+	ae_int16x4 *in = audio_stream_wrap(source, src + ioffset);
+	ae_int32x2 *out = audio_stream_wrap(sink, dst + ooffset);
+
+	while (left_samples) {
```

Comment:
all of these could be `for (left_samples = samples; left_samples; left_samples -= n)`